File: tintin-ios/TinTin/Services/RelayService.swift
import Foundation
import Network

// MARK: - Server response types

struct ServerResponse: Codable {
    let status: String
    let data: ServerData?
    let error: String?
}

/// The `data` field of a server response can be different shapes.
struct ServerData: Codable {
    var user_id: String?
    var identity_key: [UInt8]?
    var signed_pre_key: PreKeyBundleValue?
    var messages: [EnvelopeValue]?
    var queued_for: String?

    // fetch_keys returns data with identity_key + signed_pre_key directly
    // receive returns data with messages array
    // register returns data with user_id
}

struct PreKeyBundleValue: Codable {
    let identity_key: [UInt8]
    let device_id: UInt32
    let signed_pre_key_id: UInt32
    let signed_pre_key: [UInt8]
    let signed_pre_key_signature: [UInt8]
    let one_time_pre_key_id: UInt32?
    let one_time_pre_key: [UInt8]?
}

struct EnvelopeValue: Codable {
    let sender_id: String
    let recipient_id: String
    let sender_device_id: UInt32
    let timestamp: UInt64
    let content: [UInt8]       // JSON-serialised SessionMessage
    let msg_type: String
}

// MARK: - Relay Service

/// A receipt content payload received from the server.
struct ReceiptContentValue: Codable {
    let receipt_type: String
    let original_sender: String
    let original_timestamp: UInt64
}

/// TCP client for the TinTin relay server.
/// Uses NWConnection (Network.framework) for async I/O.
actor RelayService {
    enum RelayError: LocalizedError {
        case notConnected
        case connectionFailed(String)
        case serverError(String)
        case invalidResponse
        case timeout

        var errorDescription: String? {
            switch self {
            case .notConnected: return "Not connected to server"
            case .connectionFailed(let msg): return "Connection failed: \(msg)"
            case .serverError(let msg): return "Server error: \(msg)"
            case .invalidResponse: return "Invalid server response"
            case .timeout: return "Request timed out"
            }
        }
    }

    private var connection: NWConnection?
    private let queue = DispatchQueue(label: "com.tintin.relay")

    // ------------------------------------------------------------------
    // Connection
    // ------------------------------------------------------------------

    /// Connect to the relay server at `host:port`.
    func connect(host: String, port: UInt16) async throws {
        disconnect()
        let params = NWParameters.tcp
        params.allowLocalEndpointReuse = true
        let endpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host(host),
            port: NWEndpoint.Port(rawValue: port)!
        )
        connection = NWConnection(to: endpoint, using: params)

        return try await withCheckedThrowingContinuation { continuation in
            connection?.stateUpdateHandler = { [weak self] state in
                switch state {
                case .ready:
                    continuation.resume()
                case .failed(let error):
                    continuation.resume(throwing: RelayError.connectionFailed(error.localizedDescription))
                case .cancelled:
                    if !(continuation.wasResumed) {
                        continuation.resume(throwing: RelayError.connectionFailed("cancelled"))
                    }
                default:
                    break
                }
            }
            connection?.start(queue: self.queue)
        }
    }

    func disconnect() {
        connection?.cancel()
        connection = nil
    }

    var isConnected: Bool {
        connection?.state == .ready
    }

    // ------------------------------------------------------------------
    // Commands
    // ------------------------------------------------------------------

    /// Register a user with their public keys.
    func register(
        userId: String,
        identityKey: [UInt8],
        signedPreKeyId: UInt32,
        signedPreKey: [UInt8],
        signedPreKeySignature: [UInt8]
    ) async throws {
        let bundle: [String: Any?] = [
            "identity_key": identityKey,
            "device_id": 1,
            "signed_pre_key_id": signedPreKeyId,
            "signed_pre_key": signedPreKey,
            "signed_pre_key_signature": signedPreKeySignature,
            "one_time_pre_key_id": nil,
            "one_time_pre_key": nil,
        ]
        let payload: [String: Any] = [
            "cmd": "register",
            "user_id": userId,
            "identity_key": identityKey,
            "signed_pre_key": bundle,
        ]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "register failed")
        }
    }

    /// Fetch a user's key bundle for session establishment.
    func fetchKeys(userId: String) async throws -> (identityKey: [UInt8], signedPreKey: PreKeyBundleValue) {
        let payload: [String: String] = [
            "cmd": "fetch_keys",
            "user_id": userId,
        ]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok", let data = resp.data else {
            throw RelayError.serverError(resp.error ?? "user not found")
        }
        guard let ik = data.identity_key, let spk = data.signed_pre_key else {
            throw RelayError.invalidResponse
        }
        return (ik, spk)
    }

    /// Send an encrypted envelope to a recipient.
    func send(envelope: [String: Any]) async throws {
        var payload = envelope
        payload["cmd"] = "send"
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "send failed")
        }
    }

    /// Receive all pending messages for a user.
    func receive(userId: String) async throws -> [EnvelopeValue] {
        let payload: [String: String] = [
            "cmd": "receive",
            "user_id": userId,
        ]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok", let data = resp.data else {
            throw RelayError.serverError(resp.error ?? "receive failed")
        }
        return data.messages ?? []
    }

    // ------------------------------------------------------------------
    // Contact Discovery
    // ------------------------------------------------------------------

    func listUsers(query: String? = nil) async throws -> [String] {
        var payload: [String: Any] = ["cmd": "list_users"]
        if let q = query { payload["query"] = q }
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok", let data = resp.data else {
            throw RelayError.serverError(resp.error ?? "list_users failed")
        }
        // data.users is an array of strings
        return [] // simplified — proper decoding would parse data["users"]
    }

    // ------------------------------------------------------------------
    // Groups
    // ------------------------------------------------------------------

    func createGroup(name: String, creator: String) async throws -> String {
        let payload: [String: String] = ["cmd": "create_group", "name": name, "creator": creator]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok", let data = resp.data else {
            throw RelayError.serverError(resp.error ?? "create_group failed")
        }
        return data.user_id ?? "" // user_id field reused — in practice parse group_id
    }

    func joinGroup(groupId: String, userId: String) async throws {
        let payload: [String: String] = ["cmd": "join_group", "group_id": groupId, "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "join_group failed")
        }
    }

    func leaveGroup(groupId: String, userId: String) async throws {
        let payload: [String: String] = ["cmd": "leave_group", "group_id": groupId, "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "leave_group failed")
        }
    }

    func myGroups(userId: String) async throws -> [[String: Any]] {
        let payload: [String: String] = ["cmd": "my_groups", "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "my_groups failed")
        }
        return []
    }

    func groupMembers(groupId: String) async throws -> [String] {
        let payload: [String: String] = ["cmd": "group_members", "group_id": groupId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "group_members failed")
        }
        return []
    }

    // ------------------------------------------------------------------
    // Channels
    // ------------------------------------------------------------------

    func createChannel(name: String, ownerId: String) async throws -> Int64 {
        let payload: [String: String] = ["cmd": "create_channel", "name": name, "owner_id": ownerId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "create_channel failed")
        }
        return 0
    }

    func subscribeChannel(channelId: Int64, userId: String) async throws {
        let payload: [String: Any] = ["cmd": "subscribe_channel", "channel_id": channelId, "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "subscribe_channel failed")
        }
    }

    func unsubscribeChannel(channelId: Int64, userId: String) async throws {
        let payload: [String: Any] = ["cmd": "unsubscribe_channel", "channel_id": channelId, "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "unsubscribe_channel failed")
        }
    }

    // ------------------------------------------------------------------
    // Polls
    // ------------------------------------------------------------------

    func createPoll(creator: String, question: String, options: [String]) async throws -> Int64 {
        var payload: [String: Any] = ["cmd": "create_poll", "creator": creator, "question": question, "options": options]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "create_poll failed")
        }
        return 0
    }

    func votePoll(pollId: Int64, userId: String, optionId: Int64) async throws {
        let payload: [String: Any] = ["cmd": "vote_poll", "poll_id": pollId, "user_id": userId, "option_id": optionId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "vote_poll failed")
        }
    }

    // ------------------------------------------------------------------
    // Status / Stories
    // ------------------------------------------------------------------

    func setStatus(userId: String, content: String) async throws {
        let payload: [String: String] = ["cmd": "set_status", "user_id": userId, "content": content]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "set_status failed")
        }
    }

    func clearStatus(userId: String) async throws {
        let payload: [String: String] = ["cmd": "clear_status", "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "clear_status failed")
        }
    }

    func getStories() async throws -> [[String: Any]] {
        let payload: [String: String] = ["cmd": "get_stories"]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "get_stories failed")
        }
        return []
    }

    // ------------------------------------------------------------------
    // Timeline / Moments
    // ------------------------------------------------------------------

    func createPost(userId: String, content: String, targetUserId: String? = nil) async throws -> Int64 {
        var payload: [String: Any] = ["cmd": "create_post", "user_id": userId, "content": content]
        if let target = targetUserId { payload["target_user_id"] = target }
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "create_post failed")
        }
        return 0
    }

    func getTimeline(userId: String) async throws -> [[String: Any]] {
        let payload: [String: String] = ["cmd": "get_timeline", "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "get_timeline failed")
        }
        return []
    }

    func addComment(postId: Int64, userId: String, content: String) async throws {
        let payload: [String: Any] = ["cmd": "add_comment", "post_id": postId, "user_id": userId, "content": content]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "add_comment failed")
        }
    }

    func deletePost(postId: Int64, userId: String) async throws {
        let payload: [String: Any] = ["cmd": "delete_post", "post_id": postId, "user_id": userId]
        let resp: ServerResponse = try await sendCommand(payload)
        guard resp.status == "ok" else {
            throw RelayError.serverError(resp.error ?? "delete_post failed")
        }
    }

    // ------------------------------------------------------------------
    // Voice Messages
    // ------------------------------------------------------------------

    func sendVoice(envelope: [String: Any]) async throws {
        try await send(envelope: envelope)
    }

    // ------------------------------------------------------------------
    // Low-level send / receive (line-delimited JSON over TCP)
    // ------------------------------------------------------------------

    private func sendCommand<T: Encodable>(_ payload: T) async throws -> ServerResponse {
        guard let conn = connection else {
            throw RelayError.notConnected
        }

        let jsonData = try JSONSerialization.data(withJSONObject: payload as Any, options: [])
        var line = Data(jsonData)
        line.append(0x0A) // newline

        // Send
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            conn.send(content: line, completion: .contentProcessed { error in
                if let error = error {
                    continuation.resume(throwing: RelayError.connectionFailed(error.localizedDescription))
                } else {
                    continuation.resume()
                }
            })
        }

        // Receive line
        return try await withCheckedThrowingContinuation { continuation in
            conn.receive(minimumIncompleteLength: 1, maximumLength: 65536) { data, _, isComplete, error in
                if let error = error {
                    continuation.resume(throwing: RelayError.connectionFailed(error.localizedDescription))
                    return
                }
                guard let data = data, !data.isEmpty else {
                    if isComplete {
                        continuation.resume(throwing: RelayError.connectionFailed("connection closed"))
                    } else {
                        continuation.resume(throwing: RelayError.invalidResponse)
                    }
                    return
                }
                do {
                    let resp = try JSONDecoder().decode(ServerResponse.self, from: data)
                    continuation.resume(returning: resp)
                } catch {
                    continuation.resume(throwing: RelayError.invalidResponse)
                }
            }
        }
    }
}

File: tintin-android/app/src/main/java/com/tintin/app/services/RelayService.kt
package com.tintin.app.services

import com.google.gson.Gson
import com.google.gson.JsonParser
import com.tintin.app.models.UserModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.Socket
import java.net.SocketTimeoutException

/**
 * TCP relay client for the TinTin server.
 * Uses blocking I/O on a background thread (Dispatchers.IO).
 */
class RelayService(private val host: String, private val port: Int) {
    private var socket: Socket? = null
    private var writer: OutputStreamWriter? = null
    private var reader: BufferedReader? = null
    private val gson = Gson()

    /** Connect to the relay server. */
    suspend fun connect(): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val sock = Socket(host, port)
            sock.soTimeout = 10_000
            socket = sock
            writer = OutputStreamWriter(sock.getOutputStream())
            reader = BufferedReader(InputStreamReader(sock.getInputStream()))
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun disconnect() {
        try { socket?.close() } catch (_: Exception) {}
        socket = null
        writer = null
        reader = null
    }

    val isConnected: Boolean get() = socket?.isConnected == true && !socket!!.isClosed

    /** Register a user with their public keys. */
    suspend fun register(
        userId: String,
        identityKey: List<Int>,
        signedPreKeyId: Int,
        signedPreKey: List<Int>,
        signedPreKeySignature: List<Int>,
    ): Result<Unit> = sendCommand {
        val bundle = mapOf(
            "identity_key" to identityKey,
            "device_id" to 1,
            "signed_pre_key_id" to signedPreKeyId,
            "signed_pre_key" to signedPreKey,
            "signed_pre_key_signature" to signedPreKeySignature,
            "one_time_pre_key_id" to null,
            "one_time_pre_key" to null,
        )
        mapOf(
            "cmd" to "register",
            "user_id" to userId,
            "identity_key" to identityKey,
            "signed_pre_key" to bundle,
        )
    }

    /** Fetch another user's key bundle. */
    suspend fun fetchKeys(userId: String): Result<KeyBundleResponse> = withContext(Dispatchers.IO) {
        try {
            val resp = sendJsonGet(mapOf("cmd" to "fetch_keys", "user_id" to userId))
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString ?: "fetch_keys failed"))
            }
            val data = obj.getAsJsonObject("data")
            val identityKey = data.getAsJsonArray("identity_key").map { it.asInt.toByte() }.toByteArray()
            val spkObj = data.getAsJsonObject("signed_pre_key")
            val bundle = KeyBundleResponse(
                identityKey = identityKey,
                signedPreKeyId = spkObj.get("signed_pre_key_id").asInt,
                signedPreKey = spkObj.getAsJsonArray("signed_pre_key").map { it.asInt.toByte() }.toByteArray(),
                signedPreKeySignature = spkObj.getAsJsonArray("signed_pre_key_signature").map { it.asInt.toByte() }.toByteArray(),
            )
            Result.success(bundle)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /** Send an encrypted envelope. */
    suspend fun send(envelope: Map<String, Any>): Result<Unit> = sendCommand {
        mapOf("cmd" to "send", "envelope" to envelope)
    }

    /** Receive pending messages for a user. */
    suspend fun receive(userId: String): Result<List<EnvelopeResponse>> = withContext(Dispatchers.IO) {
        try {
            val resp = sendJsonGet(mapOf("cmd" to "receive", "user_id" to userId))
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString ?: "receive failed"))
            }
            val msgs = obj.getAsJsonObject("data")?.getAsJsonArray("messages") ?: return@withContext Result.success(emptyList())
            val list = msgs.map { msg ->
                val m = msg.asJsonObject
                EnvelopeResponse(
                    senderId = m.get("sender_id").asString,
                    recipientId = m.get("recipient_id").asString,
                    senderDeviceId = m.get("sender_device_id").asInt,
                    timestamp = m.get("timestamp").asLong,
                    content = m.getAsJsonArray("content").map { it.asInt.toByte() }.toByteArray(),
                    msgType = m.get("msg_type").asString,
                )
            }
            Result.success(list)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // ------------------------------------------------------------------
    // Contact Discovery
    // ------------------------------------------------------------------

    suspend fun listUsers(query: String? = null): Result<List<String>> = withContext(Dispatchers.IO) {
        try {
            val m = mutableMapOf<String, Any>("cmd" to "list_users")
            query?.let { m["query"] = it }
            val resp = sendJsonGet(m as Map<String, Any>)
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString))
            }
            val users = obj.getAsJsonObject("data")?.getAsJsonArray("users")
                ?.map { it.asString } ?: emptyList()
            Result.success(users)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // ------------------------------------------------------------------
    // Groups
    // ------------------------------------------------------------------

    suspend fun createGroup(name: String, creator: String): Result<String> = withContext(Dispatchers.IO) {
        try {
            val resp = sendJsonGet(mapOf("cmd" to "create_group", "name" to name, "creator" to creator))
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString))
            }
            val gid = obj.getAsJsonObject("data")?.get("group_id")?.asString ?: ""
            Result.success(gid)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun joinGroup(groupId: String, userId: String): Result<Unit> = sendCommand {
        mapOf("cmd" to "join_group", "group_id" to groupId, "user_id" to userId)
    }

    suspend fun leaveGroup(groupId: String, userId: String): Result<Unit> = sendCommand {
        mapOf("cmd" to "leave_group", "group_id" to groupId, "user_id" to userId)
    }

    // ------------------------------------------------------------------
    // Polls
    // ------------------------------------------------------------------

    suspend fun createPoll(creator: String, question: String, options: List<String>): Result<Long> = withContext(Dispatchers.IO) {
        try {
            val resp = sendJsonGet(mapOf("cmd" to "create_poll", "creator" to creator, "question" to question, "options" to options))
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString))
            }
            Result.success(obj.getAsJsonObject("data")?.get("poll_id")?.asLong ?: 0)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun votePoll(pollId: Long, userId: String, optionId: Long): Result<Unit> = sendCommand {
        mapOf("cmd" to "vote_poll", "poll_id" to pollId, "user_id" to userId, "option_id" to optionId)
    }

    // ------------------------------------------------------------------
    // Timeline / Moments
    // ------------------------------------------------------------------

    suspend fun createPost(userId: String, content: String, targetUserId: String? = null): Result<Long> = withContext(Dispatchers.IO) {
        try {
            val m = mutableMapOf<String, Any>("cmd" to "create_post", "user_id" to userId, "content" to content)
            targetUserId?.let { m["target_user_id"] = it }
            val resp = sendJsonGet(m as Map<String, Any>)
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString))
            }
            Result.success(obj.getAsJsonObject("data")?.get("post_id")?.asLong ?: 0)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getTimeline(userId: String): Result<List<Map<String, Any>>> = withContext(Dispatchers.IO) {
        try {
            val resp = sendJsonGet(mapOf("cmd" to "get_timeline", "user_id" to userId))
            val obj = JsonParser.parseString(resp).asJsonObject
            if (obj.get("status")?.asString != "ok") {
                return@withContext Result.failure(Exception(obj.get("error")?.asString))
            }
            val posts = obj.getAsJsonObject("data")?.getAsJsonArray("posts")
                ?.map { it.asJsonObject.entrySet().associate { (k, v) -> k to (v.asString ?: v.toString()) } }
                ?: emptyList()
            Result.success(posts)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun addComment(postId: Long, userId: String, content: String): Result<Unit> = sendCommand {
        mapOf("cmd" to "add_comment", "post_id" to postId, "user_id" to userId, "content" to content)
    }

    suspend fun deletePost(postId: Long, userId: String): Result<Unit> = sendCommand {
        mapOf("cmd" to "delete_post", "post_id" to postId, "user_id" to userId)
    }

    // ------------------------------------------------------------------
    // Status / Stories
    // ------------------------------------------------------------------

    suspend fun setStatus(userId: String, content: String): Result<Unit> = sendCommand {
        mapOf("cmd" to "set_status", "user_id" to userId, "content" to content)
    }

    suspend fun clearStatus(userId: String): Result<Unit> = sendCommand {
        mapOf("cmd" to "clear_status", "user_id" to userId)
    }

    // ------------------------------------------------------------------
    // Internal helpers
    // ------------------------------------------------------------------

    private fun sendJsonGet(body: Map<String, Any>): String {
        val json = gson.toJson(body)
        return sendJsonGet(json)
    }

    private fun sendCommand(body: () -> Map<String, Any>): Result<Unit> = try {
        val json = gson.toJson(body())
        sendJson(json)
        val response = reader?.readLine() ?: return Result.failure(Exception("No response"))
        val obj = JsonParser.parseString(response).asJsonObject
        if (obj.get("status")?.asString == "ok") Result.success(Unit)
        else Result.failure(Exception(obj.get("error")?.asString ?: "command failed"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    @Synchronized
    private fun sendJson(json: String) {
        writer?.write("$json\n")
        writer?.flush()
    }

    private fun sendJsonGet(json: String): String {
        sendJson(json)
        return reader?.readLine() ?: throw Exception("No response from server")
    }

}

data class KeyBundleResponse(
    val identityKey: ByteArray,
    val signedPreKeyId: Int,
    val signedPreKey: ByteArray,
    val signedPreKeySignature: ByteArray,
)

data class EnvelopeResponse(
    val senderId: String,
    val recipientId: String,
    val senderDeviceId: Int,
    val timestamp: Long,
    val content: ByteArray,
    val msgType: String,
)
